tests/functional/conftest.py

```diff
@@ -0,0 +1,268 @@
+    assert (
+        check_call(f"sudo mv {temp_config_file} {snap_config_file}".split())
+        == 0  # noqa
+    )
+
+    assert os.path.isfile(snap_config_file)
+
+
+def get_snap_version():
+    """Get the version of the installed snap."""
+    try:
+        output = (
+            check_output(f"snap info {SNAP_NAME}".split()).decode().strip().splitlines()
+        )
+        package_info = [i for i in output if "installed:" in i]
+        version = package_info[0].split("(")[1].split(")")[0]
+    except Exception as e:
+        logging.error("Error getting snap version: %s", e)
+
+    return version
+
+
+@pytest.fixture(scope="session", autouse=True)
+def setup_snap():
+    """Install the package to the system and cleanup afterwards.
+
+    An environment variable TEST_SNAP is needed to install the snap.
+    """
+    test_snap = os.environ.get("TEST_SNAP", None)
+    if test_snap:
+        logging.info("Installing %s snap package...", test_snap)
+        assert os.path.isfile(test_snap)
+        assert (
+            check_call(f"sudo snap install --dangerous {test_snap}".split())
+            == 0  # noqa
+        )
+
+        version = get_snap_version()
+        configure_snap(version)
+
+        assert check_call(f"sudo snap start --enable {SNAP_NAME}".split()) == 0  # noqa
+    else:
+        logging.error(
+            "Could not find %s snap package for testing. Needs to build it first.",
+            SNAP_NAME,
+        )
+
+    yield test_snap
+
+    if test_snap:
+        logging.info("Removing %s snap package...", SNAP_NAME)
+        check_call(f"sudo snap stop --disable {SNAP_NAME}".split())
+        check_call(f"sudo snap remove {SNAP_NAME}".split())
+
+
+@pytest.fixture(scope="session", autouse=True)
+def setup_test_model():
+    """Create a test model for machine manipulation."""
+    test_model_name = "prometheus-juju-exporter-func-test"
+    logging.info("Creating model '%s'...", test_model_name)
+    assert check_call(f"juju add-model {test_model_name}".split()) == 0  # noqa
+
+    yield test_model_name
+
+    logging.info("Destroying model '%s'...", test_model_name)
+    check_call(
+        f"juju destroy-model --destroy-storage --force -y {test_model_name}".split()
+    )
+
+
+@pytest.fixture
+def get_registry_data():
+    """Connect to the registry and wait to get labelbalues data upon completing the collection."""
+
+    def _get_registry():
+        previous_registry_data = None
+        registry_data = None
+        connection_retry = 0
+
+        # Make sure the registry_data is accessible and collection process
+        # is complete (no more labelvalues added to registry)
+        while connection_retry < 10 and (
+            registry_data is None or registry_data != previous_registry_data
+        ):
+            try:
+                previous_registry_data = registry_data
+                localhost = urllib.request.urlopen("http://localhost:5000/", timeout=60)
```

Comment:
Is the hardcoded URL guaranteed to work?

Reply:
Sine the port is configured within the functional tests, the url *should* always work. But to make it looks less "hard-coded", I've set a global variable for the port.